qiskit/synthesis/clifford/clifford_decompose_layers.py

```diff
@@ -0,0 +1,392 @@
+        if (stabz_update != stabz_update.T).any():
+            raise QiskitError(
+                "The multiplication of stabx_inv and stab_z is not a symmetric matrix."
+            )
+
+    CZ1_circ = cz_synth_func(stabz_update, validate=validate)
+
+    for j in range(num_qubits):
+        for i in range(0, j):
+            if stabz_update[i][j]:
+                _append_cz(cliff_cpy, i, j)
+
+    for i in range(0, num_qubits):
+        if stabz_update[i][i]:
+            S1_circ.s(i)
+            _append_s(cliff_cpy, i)
+
+    for qubit in range(num_qubits):
+        H2_circ.h(qubit)
+        _append_h(cliff_cpy, qubit)
+
+    return H2_circ, CZ1_circ, S1_circ, cliff_cpy
+
+
+def _decompose_hadamard_free(cliff, validate, cz_synth_func, cx_synth_func, cx_cz_synth_func):
+    """Assumes that the Clifford cliff is Hadamard free.
+    Decompose it into the layers S2 - CZ2 - CX, where
+    S2_circ is a circuit that can contain only S gates,
+    CZ2_circ is a circuit that can contain only CZ gates, and
+    CX_circ is a circuit that can contain CX gates.
+
+    Args:
+        cliff (Clifford): a Hadamard-free clifford operator.
+        validate (Boolean): if True, validates the synthesis process.
+        cz_synth_func (Callable): a function to decompose the CZ sub-circuit.
+        cx_synth_func (Callable): a function to decompose the CX sub-circuit.
+        cx_cz_synth_func (Callable): optional, a function to decompose both sub-circuits CZ and CX.
+
+    Return:
+        S2_circ: a circuit that can contain only S gates.
+        CZ2_circ: a circuit that can contain only CZ gates.
+        CX_circ: a circuit that can contain only CX gates.
+
+    Raises:
+        QiskitError: if cliff is not Hadamard free.
+    """
+
+    num_qubits = cliff.num_qubits
+    destabx = cliff.destab_x
+    destabz = cliff.destab_z
+    stabx = cliff.stab_x
+
+    if not (stabx == np.zeros((num_qubits, num_qubits))).all():
+        raise QiskitError("The given Clifford is not Hadamard-free.")
+
+    destabz_update = np.matmul(calc_inverse_matrix(destabx), destabz) % 2
+    # Assert that destabz_update is a symmetric matrix.
+    if validate:
+        if (destabz_update != destabz_update.T).any():
+            raise QiskitError(
+                "The multiplication of the inverse of destabx and"
+                "destabz is not a symmetric matrix."
+            )
+
+    S2_circ = QuantumCircuit(num_qubits, name="S2")
+    for i in range(0, num_qubits):
+        if destabz_update[i][i]:
+            S2_circ.s(i)
+
+    if cx_cz_synth_func is not None:
+        CZ2_circ, CX_circ = cx_cz_synth_func(
+            destabz_update, cliff.destab_x.transpose(), num_qubits=num_qubits
+        )
+        return S2_circ, CZ2_circ, CX_circ
+
+    CZ2_circ = cz_synth_func(destabz_update, validate=validate)
+
+    mat = destabx.transpose()
+    CX_circ = cx_synth_func(mat, validate=validate)
+
+    return S2_circ, CZ2_circ, CX_circ
+
+
+def _fix_pauli(cliff, cliff_target):
```

Comment:
The name of the function suggests a side-effect, but it doesn't affect anything. Maybe call it `_calc_pauli_diff` or something similar